src/main/java/seedu/address/model/note/TagNoteMap.java

```diff
@@ -0,0 +1,117 @@
+package seedu.address.model.note;
+
+
+import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import seedu.address.model.client.Client;
+import seedu.address.model.tag.Tag;
+
+
+/**
+ * Manages the relationship between Tags and Notes.
+ */
+public class TagNoteMap {
+    //    TODO: Add implementation
+    public final Set<Tag> tagSet = new HashSet<>(); // probably redundant todo: remove later
+    public final Set<Note> noteSet = new HashSet<>();
+    public final Map<Tag, Set<Note>> tagToNotesMap = new HashMap<>();
+    public final Map<Note, Set<Tag>> noteToTagsMap = new HashMap<>();
+
+
+    /**
+     * Initialises the TagNoteMap from a list of clients.
+     *
+     * @param clients The list of clients, each containing their notes and associated tags.
+     */
+    public void initTagNoteMapFromClients(List<Client> clients) {
+        requireAllNonNull(clients);
+        for (Client client : clients) {
+            Set<Note> clientNotes = client.getClientNotes();
+            noteSet.addAll(clientNotes);
+            for (Note clientNote : clientNotes) {
+                Set<Tag> tags = clientNote.getTags();
+                tagSet.addAll(tags);
+                updateTagsForNote(tags, clientNote);
+            }
+        }
+        // todo: how to remove old entries to the map?
+    }
+
+    /**
+     * Initialises the TagNoteMap from a list of countries.
+     *
+     * @param countryNotes The set of countries, each containing their notes and associated tags.
+     */
+    public void initTagNoteMapFromCountryNotes(Set<Note> countryNotes) {
+        // todo: make init work when passed in a list of countryNotes
+        requireAllNonNull(countryNotes);
+        noteSet.addAll(countryNotes);
+        for (Note countryNote : countryNotes) {
+            Set<Tag> tags = countryNote.getTags();
+            tagSet.addAll(tags);
+            updateTagsForNote(tags, countryNote);
+        }
+    }
+
+    // todo: how to remove old entries to the map? <== need to do this to "sync"
```

Comment:
I'm not sure what you mean. Can you give an example of this?

Reply:
this is a moot point for now, it's relevant when deleting/updating a note and the TagNoteMap has to be updated